File: app/src/main/java/com/minar/randomix/utilities/ShakeEventListener.java
package com.minar.randomix.utilities;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;

// A listener to detect a shake gesture
public class ShakeEventListener implements SensorEventListener {
    // Minimum movement force
    private static final int MIN_FORCE = 14;
    // Number of direction changes
    private static final int MIN_DIRECTION_CHANGE = 4;
    // Maximum pause between movements
    private static final int MAX_PAUSE_BETWEEN_DIRECTION_CHANGE = 200;
    // Maximum time for the whole shake gesture
    private static final int MAX_TOTAL_DURATION_OF_SHAKE = 500;
    // Time for gesture start, last movement start and number of movements
    private long mFirstDirectionChangeTime = 0;
    private long mLastDirectionChangeTime;
    private int mDirectionChangeCount = 0;

    // Last x, y and z positions
    private float lastX = 0;
    private float lastY = 0;
    private float lastZ = 0;

    // Listener called when a shake is detected
    private OnShakeListener mShakeListener;

    // Interface for the listener
    public interface OnShakeListener {
        void onShake();
    }

    // Set the listener
    public void setOnShakeListener(OnShakeListener listener) {
        mShakeListener = listener;
    }

    public void onSensorChanged(SensorEvent se) {
        // Get sensor data (x, y, z)
        float x = se.values[0];
        float y = se.values[1];
        float z = se.values[2];

        // Calculate movement
        float totalMovement = Math.abs(x + y + z - lastX - lastY - lastZ);

        if (totalMovement > MIN_FORCE) {
            // Get time
            long now = System.currentTimeMillis();
            // Store first movement time
            if (mFirstDirectionChangeTime == 0) {
                mFirstDirectionChangeTime = now;
                mLastDirectionChangeTime = now;
            }

            // Check if the last movement was not long ago
            long lastChangeWasAgo = now - mLastDirectionChangeTime;
            if (lastChangeWasAgo < MAX_PAUSE_BETWEEN_DIRECTION_CHANGE) {
                // Store movement data
                mLastDirectionChangeTime = now;
                mDirectionChangeCount++;
                // Store last sensor data
                lastX = x;
                lastY = y;
                lastZ = z;
                // Check how many movements are so far
                if (mDirectionChangeCount >= MIN_DIRECTION_CHANGE) {
                    // Check total duration
                    long totalDuration = now - mFirstDirectionChangeTime;
                    if (totalDuration < MAX_TOTAL_DURATION_OF_SHAKE) {
                        mShakeListener.onShake();
                        resetShakeParameters();
                    }
                }
            } else {
                resetShakeParameters();
            }
        }
    }

    // Reset the parameters
    private void resetShakeParameters() {
        mFirstDirectionChangeTime = 0;
        mDirectionChangeCount = 0;
        mLastDirectionChangeTime = 0;
        lastX = 0;
        lastY = 0;
        lastZ = 0;
    }

    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }
}

File: app/src/main/java/com/minar/randomix/utilities/ShakeEventListener.kt
package com.minar.randomix.utilities

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import kotlin.math.abs

class ShakeEventListener : SensorEventListener {

    private val minForce = 14
    private val minDirectionChange = 4
    private val maxPauseBetweenDirectionChange = 200L
    private val maxTotalDurationOfShake = 500L

    private var firstDirectionChangeTime = 0L
    private var lastDirectionChangeTime = 0L
    private var directionChangeCount = 0

    private var lastX = 0f
    private var lastY = 0f
    private var lastZ = 0f

    private var shakeListener: OnShakeListener? = null

    fun interface OnShakeListener {
        fun onShake()
    }

    fun setOnShakeListener(listener: OnShakeListener) {
        shakeListener = listener
    }

    override fun onSensorChanged(se: SensorEvent) {
        val x = se.values[0]
        val y = se.values[1]
        val z = se.values[2]

        val totalMovement = abs(x + y + z - lastX - lastY - lastZ)

        if (totalMovement > minForce) {
            val now = System.currentTimeMillis()
            if (firstDirectionChangeTime == 0L) {
                firstDirectionChangeTime = now
                lastDirectionChangeTime = now
            }
            val lastChangeWasAgo = now - lastDirectionChangeTime
            if (lastChangeWasAgo < maxPauseBetweenDirectionChange) {
                lastDirectionChangeTime = now
                directionChangeCount++
                lastX = x; lastY = y; lastZ = z
                if (directionChangeCount >= minDirectionChange) {
                    val totalDuration = now - firstDirectionChangeTime
                    if (totalDuration < maxTotalDurationOfShake) {
                        shakeListener?.onShake()
                        resetShakeParameters()
                    }
                }
            } else {
                resetShakeParameters()
            }
        }
    }

    private fun resetShakeParameters() {
        firstDirectionChangeTime = 0L
        directionChangeCount = 0
        lastDirectionChangeTime = 0L
        lastX = 0f; lastY = 0f; lastZ = 0f
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) = Unit
}
